Python/Product/Analysis/AnalysisHashSet.cs

```diff
@@ -228,23 +268,41 @@ public AnalysisHashSet(IEnumerable<AnalysisValue> enumerable, IEqualityComparer<
             // anything other than checking equality, as it does
             // not have the merge logic we need.
             var seen = new HashSet<AnalysisValue>(Comparer);
-            foreach (var b in buckets) {
+            foreach (var b in buckets.Buckets) {
                 if (b.Key != null && b.Key != _removed && !seen.Add(b.Key)) {
                     // Cannot reuse the buckets
                     return new AnalysisHashSet(Comparer).Union(this, canMutate: true);
                 }
             }
 
             var res = new AnalysisHashSet(Comparer);
-            var newBuckets = new Bucket[buckets.Length];
-            Array.Copy(buckets, newBuckets, buckets.Length);
+            var newBuckets = new BucketSet(buckets.Capacity, buckets.Count);
+            Array.Copy(buckets.Buckets, newBuckets.Buckets, buckets.Capacity);
 
-            res._buckets = newBuckets;
-            res._count = count;
+            res.Buckets = newBuckets;
             return res;
         }
 
         public bool SetEquals(IAnalysisSet other) {
+            if (ReferenceEquals(this, other)) {
+                return true;
+            }
+
+            if (Comparer == other.Comparer) {
```

Comment:
What is the scenario when comparers aren't equal?

Reply:
I don't understand the question? We have a selection of 4-5 different comparers that affect how values are combined, so it's very easy for two different sets to have different comparers.

Reply:
Nevermind, I got it.